src/speed-index.js

```diff
@@ -33,14 +33,44 @@ function calculateFrameProgress(current, initial, target) {
 	return progress;
 }
 
-function calculateVisualProgress(frames) {
+function calculateProgressBetweenFrames(frames, lowerBound, upperBound, threshold, getProgress, setProgress) {
+	if (threshold === 0) {
+		frames.forEach(frame => setProgress(frame, getProgress(frame)));
+		return;
+	}
+
+	const lowerFrame = frames[lowerBound];
+	const upperFrame = frames[upperBound];
+
+	const lowerProgress = getProgress(lowerFrame);
+	const upperProgress = getProgress(upperFrame);
+
+	setProgress(lowerFrame, lowerProgress);
+	setProgress(upperFrame, upperProgress);
+
+	if (Math.abs(lowerProgress - upperProgress) < threshold) {
+		for (let i = lowerBound; i < upperBound; i++) {
+			setProgress(frames[i], lowerProgress);
+		}
+	} else if (upperBound - lowerBound > 1) {
+		const midpoint = Math.floor((lowerBound + upperBound) / 2);
+		calculateProgressBetweenFrames(frames, lowerBound, midpoint, threshold, getProgress, setProgress);
+		calculateProgressBetweenFrames(frames, midpoint, upperBound, threshold, getProgress, setProgress);
+	}
+}
+
+function calculateVisualProgress(frames, opts) {
 	const initial = frames[0];
 	const target = frames[frames.length - 1];
 
-	frames.forEach(function (frame) {
-		const progress = calculateFrameProgress(frame, initial, target);
-		frame.setProgress(progress);
-	});
+	calculateProgressBetweenFrames(
+		frames,
+		0,
+		frames.length - 1,
+		opts && opts.fast ? 5 : 0,
+		frame => frame.getProgress() || calculateFrameProgress(frame, initial, target),
```

Comment:
maybe not enough coffee, but ... brain hurts. 

this is an expression of `fn1 || fn2` and will be passed in as the `getProgress` parameter. 
when called, the fn1 will evaluate and then return and if that result is falsy then the `fn2` will then run?

is there another way to model this?